### PR TITLE
Update Kubernetes Helm Chart

### LINKED ISSUE #2012: docs: Add example to demonstrate deployment on k8s along with a helm chart
This has been a heavily requested feature.

See https://github.com/cdr/code-server/issues/914

### PR DESCRIPTION
**Pull request by Matthew-Beckett**
Fixes #2012, #914

### CODE REVIEW

**Review by nhooyr — COMMENTED** [review]
I haven't looked into this hard but is it possible to simplify this? 700 additional lines is a lot of code, we don't have to serve every use case.

**Comment by Matthew-Beckett** [comment]
@nhooyr the 700 lines is generated from helm create, and as such is considered best practice. This actually only caters for a very simple use case 😅 

The original pull request for this chart was 600 lines and did not include a helm ignore which would explain why mine is so large.

https://github.com/cdr/code-server/pull/917/commits/615948c73fd9eca56fe7672a77b5b4acd04f5411

**Comment by Matthew-Beckett** [comment]
@nhooyr would you be able to advise on my reply? I can begin trying to shrink the commit if my reply is unacceptable.

**Comment by alexgorbatchev** [comment]
This is a very slim helm chart IMO. First 200 lines are `.gitgnore` and `README.md`. The only to slim this down is by removing comments.

+1 for merging this in. To make this chart useful it needs to be published via Github pages.

**Comment by nhooyr** [comment]
> @nhooyr would you be able to advise on my reply? I can begin trying to shrink the commit if my reply is unacceptable.

Nah you're good! I just need to find the time to review this/dig into helm. I don't want to publish anything we haven't tested and verified ourselves.

What's the advantage to `helm` here versus shipping a couple basic k8s manifest files? That'd be much smaller, easier to audit and understand. And just as easy to delete by targeting a label `cdr/code-server`.

**Comment by nhooyr** [comment]
Also cc @sreya92 for thoughts as I know he's been in helm land for a while.

**Comment by Matthew-Beckett** [comment]
> To make this chart useful it needs to be published via Github pages.

@alexgorbatchev absolutely, I agree! I haven't ever done this myself, however I am familiar with how it works. So that the solution is supportable and of a good quality, you mind pairing with me on this?

> What's the advantage to helm here versus shipping a couple basic k8s manifest files?

@nhooyr indeed, there is no argument the complexity of Helm charts is considerable when compared to that of native manifests.

However, generally it is accepted that Helm charts are a more reliable way of an application vendor providing a reliable mechanism for deploying the application to their cluster that is managed and life cycled. Helm charts are ultimately packages in a package manager, so it's the argument of copying a binary to /usr/local/bin over installing it through your distribution's package manager and having versioning and state management capability. 

Apologies if this is an overly dumb explanation! 😅 I'm not sure how else to explain it.

**Comment by alexgorbatchev** [comment]
> > To make this chart useful it needs to be published via Github pages.
> 
> @alexgorbatchev absolutely, I agree! I haven't ever done this myself, however I am familiar with how it works. So that the solution is supportable and of a good quality, you mind pairing with me on this?

Sure, I can help. 

> > What's the advantage to helm here versus shipping a couple basic k8s manifest files?
> 
> @nhooyr indeed, there is no argument the complexity of Helm charts is considerable when compared to that of native manifests.

The main advantage is composability. Every engineering organization uses some form of provisioning system like Terraform or Ansible to manage their infrastructure. Helm charts provide an easy way to install complex Kubernetes manifests with pre-defined configuration specific to the application. It also makes it easy to receive future updates. While you can also use plain manifest files and modify them for your needs, you will have to do this every there's an update to the manifest.

Think of a Helm chart as as a bash script that takes arguments, and manifest as a bash script that you have to heavily edit to pass the same arguments.

**Comment by Matthew-Beckett** [comment]
@alexgorbatchev how would we implement Helm repositories on GitHub pages for cdr/code-server as external contributors to the project?

**Comment by alexgorbatchev** [comment]
> @alexgorbatchev how would we implement Helm repositories on GitHub pages for cdr/code-server as external contributors to the project?

I'll have a PR soon.

**Comment by alexgorbatchev** [comment]
@Matthew-Beckett in order to get easy github pages hosting, `cdr` team needs to create a `cdr.github.io` repo and clone the repo I made as "PR" to the project:

https://github.com/alexgorbatchev/alexgorbatchev.github.io

example output: https://alexgorbatchev.github.io/helm-charts/public/

This repo includes chart files from THIS PR and assumes they will be hosted there. It makes it easier to publish github pages. 

Alternatively one could simply look at how I've implemented publishing and move it else where. Everything happens via a few simple HELM shell commands and as a bonus I'm generating `index.html` with Bash.

Please let me know if this looks sufficient.

**Comment by alexgorbatchev** [comment]
Please let me know when I can take down the example from my public page :)

**Comment by alexgorbatchev** [comment]
@code-asher @nhooyr Could you please have a look at this?

**Comment by MnrGreg** [comment]
I'm surprised how fast `code-server` starts up. Would be interesting to see how well it works with a requests based `HPA Scale to 0`. Should be able to drive quite a high density of deployments.

Just need `HPA Scale to 0` to mature. +1 here: https://github.com/kubernetes/kubernetes/issues/69687

**Comment by alexgorbatchev** [comment]
@code-asher @nhooyr ping

**Comment by Matthew-Beckett** [comment]
@code-asher @nhooyr nudge

**Comment by hammady** [comment]
@code-asher @nhooyr @Matthew-Beckett I am currently using this chart and have some comments/fixes. I can formally review this PR if it helps getting it upstream. Please let me know.

**Comment by Matthew-Beckett** [comment]
> @code-asher @nhooyr @Matthew-Beckett I am currently using this chart and have some comments/fixes. I can formally review this PR if it helps getting it upstream. Please let me know.

Absolutely! Please do! Not sure how you want to handle the merging, do you just want to PR into my feature and then subsequently into this PR?

**Comment by hammady** [comment]
> > @code-asher @nhooyr @Matthew-Beckett I am currently using this chart and have some comments/fixes. I can formally review this PR if it helps getting it upstream. Please let me know.
> 
> Absolutely! Please do! Not sure how you want to handle the merging, do you just want to PR into my feature and then subsequently into this PR?

Yeah I can do that.

**Comment by Matthew-Beckett** [comment]
> @Matthew-Beckett I have created a new PR at [Matthew-Beckett#14](https://github.com/Matthew-Beckett/code-server/pull/14).
> I will create another one for hostPath volumes support once this is merged.

@hammady will review now! Thanks so much! <3

**Comment by alexgorbatchev** [comment]
@nhooyr @code-asher please review

File: charts/code-server/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 1.0.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+appVersion: 3.5.0
```
Comment (nhooyr) [inline charts/code-server/Chart.yaml:23]: Should be `3.6.2` now!

File: charts/code-server/README.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
+# code-server
```
Comment (nhooyr) [inline charts/code-server/README.md:1]: Can you add a note that this is community maintained? Neither myself nor @code-asher are familiar with helm.

If yourself @Matthew-Beckett or @alexgorbatchev would like become the official maintainers of the helm chart, I would be happy with that as well and you can add a note here.

File: charts/code-server/README.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+![Version: 1.0.0](https://img.shields.io/badge/Version-1.0.0-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 3.5.0](https://img.shields.io/badge/AppVersion-3.5.0-informational?style=flat-square)
```
Comment (nhooyr) [inline charts/code-server/README.md:3]: `3.6.2`

File: charts/code-server/README.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+[code-server](https://github.com/cdr/code-server) code-server is VS Code running
+on a remote server, accessible through the browser.
+
+## TL;DR;
+
+```console
+$ git clone https://github.com/cdr/code-server.git
+$ helm install deployment/chart
```
Comment (nhooyr) [inline charts/code-server/README.md:12]: Will have to cd into code-server here.

File: charts/code-server/README.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+The command removes all the Kubernetes components associated with the chart and
+deletes the release.
+
+## Configuration
+
+The following table lists the configurable parameters of the nginx-ingress chart
+and their default values.
+
+
+The following table lists the configurable parameters of the code-server chart
```
Comment (nhooyr) [inline charts/code-server/README.md:56]: Duplicate line.

File: charts/code-server/README.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+Alternatively, a YAML file that specifies the values for the above parameters
+can be provided while installing the chart. For example,
+
+```console
+$ helm install --name my-release -f values.yaml deployment/chart
+```
+
+> **Tip**: You can use the default [values.yaml](values.yaml)
+
```
Comment (nhooyr) [inline charts/code-server/README.md:117]: No need for extra lines here.

**Review by nhooyr — CHANGES_REQUESTED** [review]
Sorry for the long wait in review guys! Appreciate your patience.

I've reviewed what I can and asked @sreya to help drop a review as well. Hopefully we can get this merged this week!

**Comment by Matthew-Beckett** [comment]
@nhooyr I have resolved all your comments, @alexgorbatchev if you do not wish to maintain this going forward I can remove you, it's just to get this finally merged in 😅

**Review by nhooyr — APPROVED** [review]

**Comment by nhooyr** [comment]
🎉

**Comment by Matthew-Beckett** [comment]
> 🎉

Got a bottle of champagne in the fridge I've been saving, this might be the moment 😆

**Comment by Matthew-Beckett** [comment]
> @Matthew-Beckett in order to get easy github pages hosting, `cdr` team needs to create a `cdr.github.io` (see https://pages.github.com/) repo and clone the repo I made as "PR" to the project:
> 
> https://github.com/alexgorbatchev/alexgorbatchev.github.io
> 
> example output: https://alexgorbatchev.github.io/helm-charts/public/
> 
> This repo includes chart files from THIS PR and assumes they will be hosted there. It makes it easier to publish github pages.
> 
> Alternatively one could simply look at how I've implemented publishing and move it else where. Everything happens via a few simple HELM shell commands and as a bonus I'm generating `index.html` with Bash.
> 
> The output from the `build.sh` file looks like this:
> 
> ```shell
> ❯ ./scripts/build.sh
> Packaging helm charts
> Successfully packaged chart and saved it to: public/code-server-1.0.0.tgz
> Writing: public/index.yaml
> Writing: public/index.html
> ```
> 
> Please let me know if this looks sufficient.

@nhooyr should we open this into a new issue? I think the points raised are valid and would benefit users immensely post-merge.

File: .github/CODEOWNERS [hunk 7]
```diff
@@ -1 +1,2 @@
-* @code-asher @nhooyr
+charts/code-server @Matthew-Beckett @alexgorbatchev
```
Comment (nhooyr) [inline .github/CODEOWNERS:1]: more automation is better for sure 👍🏼

**Comment by nhooyr** [comment]
I believe we can just use e.g. `https://raw.githubusercontent.com/Matthew-Beckett/code-server/feature/helm3/charts/code-server` but once it's merged it'd be `https://raw.githubusercontent.com/Matthew-Beckett/code-server/master/charts/code-server`.

Can you give it a shot and see if it works?

**Comment by nhooyr** [comment]
Ooops, nevermind I'm wrong. I see, we need to serve a `tgz` file that contains the chart.

**Comment by Matthew-Beckett** [comment]
> Ooops, nevermind I'm wrong. I see, we need to serve a `tgz` file that contains the chart.

Indeed, furthermore, Helm expects a certain HTTP response to list the charts and their version, Helm wouldn't natively support GitHub Release pulling.

**Comment by nhooyr** [comment]
Yea that's unfortunate. I'm surprised helm doesn't let you use git urls, seem's like huge mistake.

There's https://github.com/aslafy-z/helm-git at least. And now https://github.com/helm/helm/pull/6734

Personally, I'm ok with users having to clone the repository before using helm as I think it's unfortunate to have to setup github pages just for helm but I'm open to discussion.

**Comment by alexgorbatchev** [comment]
> Personally, I'm ok with users having to clone the repository before using helm as I think it's unfortunate to have to setup github pages just for helm but I'm open to discussion.

I think ideally there should be a simple way to use Helm with this chart, via github pages (or similar). Having to clone the repo here is akin to asking users to clone a repo before running `brew install` - doable, but regrettable :)

**Comment by Matthew-Beckett** [comment]
@sreya @code-asher nudge ☺️

File: charts/code-server/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        command:
+          - sh
+          - -c
+          - |
+            chown -R {{ .Values.securityContext.runAsUser }}:{{ .Values.securityContext.fsGroup }} /home/coder
+        securityContext:
+          runAsUser: {{ .Values.volumePermissions.securityContext.runAsUser }}
+        volumeMounts:
+        - name: data
+          mountPath: /home/coder
```
Comment (sreya) [inline charts/code-server/templates/deployment.yaml:45]: Do we always want the persistent volume to be mounted to `/home/coder`? If we allow for users to override the image we should maybe allow them to override the home directory.

File: charts/code-server/README.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+[code-server](https://github.com/cdr/code-server) code-server is VS Code running
+on a remote server, accessible through the browser.
+
+This chart is community maintained by [@Matthew-Beckett](https://github.com/Matthew-Beckett) and [@alexgorbatchev](https://github.com/alexgorbatchev)
+
+## TL;DR;
+
+```console
+$ git clone https://github.com/cdr/code-server.git
+$ helm install code-server/charts/code-server
```
Comment (sreya) [inline charts/code-server/README.md:14]:
```suggestion
$ helm install code-server code-server/charts/code-server
```

File: charts/code-server/values.yaml [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  ## code-server data Persistent Volume Storage Class
+  ## If defined, storageClassName: <storageClass>
+  ## If set to "-", storageClassName: "", which disables dynamic provisioning
+  ## If undefined (the default) or set to null, no storageClassName spec is
+  ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
+  ##   GKE, AWS & OpenStack)
+  ##
+  # storageClass: "-"
+  accessMode: ReadWriteOnce
+  size: 1Gi
```
Comment (sreya) [inline charts/code-server/values.yaml:115]: 1Gi seems a bit small for a default value I think 10 might be a better value.

**Review by sreya — APPROVED** [review]
Looks good overall. I obviously didn't validate all the parameters since you provided so many but a basic install works well :) If there are any bugs they will shake out over time.

**Comment by nhooyr** [comment]
Going to merge and fix comments myself so we can get this into today's release!

**Comment by nhooyr** [comment]
Thanks for reviewing @sreya

**Comment by nhooyr** [comment]
> I think ideally there should be a simple way to use Helm with this chart, via github pages (or similar). Having to clone the repo here is akin to asking users to clone a repo before running brew install - doable, but regrettable :)

I mean it's more on Helm than us though. As a project maintainer, we have to test and support so many different platforms/package managers/combinations already. Adding one more is frustrating when helm could just support git.

**Comment by nhooyr** [comment]
I opened https://github.com/cdr/code-server/issues/2302 regardless though.

**Pull request merged**